src/NodeJSService.ts

```diff
@@ -114,135 +114,36 @@ export default class LightningServiceNodeJS extends LocalMain.LightningService {
 				dialogTitle: 'Uh-oh! Local ran into an error.',
 				dialogMessage: e.toString(),
 			});
 
 			// Halt provisioning and cleanup.
-			throw new Error('error encounted during preprovision step');
+			throw new Error('error encountered during preprovision step');
 		}
 	}
 
 	async finalizeNewSite (): Promise<void> {
-		LocalMain.sendIPCEvent('updateSiteMessage', this._site.id, 'Installing headless WordPress plugins');
-		const { wpCli, siteDatabase, siteProcessManager, errorHandler } = serviceContainer.cradle;
+		const { wpCli, siteDatabase, errorHandler } = serviceContainer.cradle;
+		const { additionalPlugins = [], installCommand = '' } = this._site.customOptions;
+		const requiredPlugins = ['faustwp', 'wp-graphql', ...additionalPlugins];
 
 		try {
 			// eslint-disable-next-line default-case
 			await siteDatabase.waitForDB(this._site);
 
-			// Add GraphQL server to WordPress.
-			await wpCli.run(this._site, [
-				'plugin',
-				'install',
-				'wp-graphql',
-				'--activate',
-			]);
-
-			// Add the FaustWP plugin.
-			await wpCli.run(this._site, [
-				'plugin',
-				'install',
-				'faustwp',
-				'--activate',
-			]);
-
-			// eslint-disable-next-line @typescript-eslint/ban-ts-comment
-			// @ts-ignore
-			if (this._site.customOptions.atlasUrl) {
-				// Add the atlas-content-modeler plugin.
-				await wpCli.run(this._site, [
-					'plugin',
-					'install',
-					'atlas-content-modeler',
-					'--activate',
-				]);
-			}
-
-			// The Atlas Ecommerce template requires two extra plugins.
-			if (this._site.customOptions?.bpId === 'ecommerce-blueprint-atlas-1') {
-				/**
-				 * Install Atlas Commerce Blocks.
-				 * See https://github.com/wpengine/atlas-commerce-blocks.
-				 */
-				await wpCli.run(this._site, [
-					'plugin',
-					'install',
-					'https://wp-product-info.wpesvc.net/v1/plugins/atlas-commerce-blocks?download',
-					'--activate',
-				]);
-
-				/**
-				 * Install Atlas Commerce Connector for BigCommerce.
-				 * See https://github.com/wpengine/atlas-commerce-connector-bigcommerce.
-				 */
-				await wpCli.run(this._site, [
-					'plugin',
-					'install',
-					'https://wp-product-info.wpesvc.net/v1/plugins/atlas-commerce-connector-bigcommerce?download',
-					'--activate',
-				]);
-			}
+			LocalMain.sendIPCEvent('updateSiteMessage', this._site.id, 'Installing WordPress plugins');
 
-			// Fetch the secret key generated by the FaustWP plugin on activation.
-			const faustWPsettings = await wpCli.run(this._site, [
-				'option',
-				'get',
-				'faustwp_settings',
-				'--format=json',
-			]);
-
-			// eslint-disable-next-line @typescript-eslint/ban-ts-comment
-			// @ts-ignore
-			if (this._site.customOptions.atlasUrl) {
-				// Add the atlas-search plugin.
-				await wpCli.run(this._site, [
-					'plugin',
-					'install',
-					'atlas-search',
-					'--activate',
-				]);
+			for (const plugin of requiredPlugins) {
+				await wpCli.run(this._site, ['plugin', 'install', '--activate', plugin]);
```

Comment:
> Though maybe wpCli.run() needs to block all other plugin installs while it installs a plugin.

I think it does, unfortunately, yes — or, rather, WP-CLI itself seems to expect sequential installs.

I've tried `Promise.all()` in the past and WP-CLI wasn't happy about concurrent reads/writes to its caching folder. If I remember right it manifested as a very occasional 'plugin not found' error during the `install` attempt, with a message about non-readable/writeable `.wp-cli/cache` folder contents. (Sometimes it would work just fine, though, as with a lot of timing/file access issues.)

So I'm a little cautious about switching, as much as I'd love for it to work!

WP-CLI supports passing multiple plugins to `install` in the form `wp plugin install plugin1 plugin2`, which we could do here with something like this:

```
await wpCli.run(this._site, ['plugin', 'install', '--activate', requiredPlugins.join(' ')]);
```

But it still only installs them sequentially by looping through — it makes no attempt to parallelize the work:

https://github.com/wp-cli/extension-command/blob/17b16548b5775616dcbbd92f7836e67bba02e8ba/src/WP_CLI/CommandWithUpgrade.php#L147-L151

I also found that error messages can be less clear when passing multiple plugins to `install` (if the bulk install fails at some point, it may not always be clear which plugin caused the issue).

Reply:
@kienstra In case you're curious about the error, I switched to `await Promise.all(requiredPlugins.map …`, then created a few sites from Atlas blueprints. The first two worked fine. The third gave me this:

<img width="510" alt="Screenshot 2023-09-01 at 09 16 44" src="https://github.com/getflywheel/local-addon-atlas/assets/647669/2b69e432-4cbd-4780-a430-5590a9de795e">

So I don't know that we can safely switch to Promise.all, assuming that's the root cause of the above error. (I can't seem to trigger the same error when installing plugins sequentially.)

Reply:
> When it works it's a lot faster, but when it fails the user has to reattempt site creation from scratch.

That's a good reason to not use `Promise.all()` 😄